### ISSUE TITLE
Exception in thread Thread and hang

### ISSUE THREAD

**Issue by JaneVinWin** [Parsely/pykafka#314]
when i consumed for a while, balanced consumer will raise a exception on screen, see below:
```
No handlers could be found for logger "kazoo.client"
Exception in thread Thread-15:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 36, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 333, in _get_children
    result = self._func(children)
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 532, in _consumers_changed
    self._rebalance()
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 431, in _rebalance
    partitions = self._decide_partitions(participants)
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 320, in _decide_partitions
    idx = participants.index(self._consumer_id)
ValueError: 'localhost.localdomain:dc19b88a-7233-4f44-8295-fd196f72a2b3' is not in list
 ```

However, the loop still hang there but not exit. and this exception can not be caught in code. See below:
for message in balanced_consumer:
            
            try:
                print "offset: " + str(message.offset)
                            
            except ex.SocketDisconnectedError,e:
                a = datetime.datetime.now()
                print "SocketDisconnectedError inside at: " + str(a) + ";",e
                break
            except ValueError,e:
                a = datetime.datetime.now()
                print "ValueError inside at: " + str(a) + ";",e
                break
            except Exception, e:
                a = datetime.datetime.now()
                print Exception, " inside at: " + str(a) + ";", e
                break
            else:
                pass

**Comment by yungchin** [comment]
Thanks for reporting this. I believe this issue occurs when your zookeeper connection has been reset, which was very recently fixed in #308. Could you please check if you still see this if you use current master?

**Comment by JaneVinWin** [comment]
@yungchin , thanks for your reply, could you tell me which code i should use? i'm confusing about a lot comment in #308

**Comment by yungchin** [comment]
Hi, yes, sorry, #308 has been merged into master already, so if you just check out master you should get those changes with it. Thanks

**Comment by JaneVinWin** [comment]
@yungchin i checkout master code to run. it could run normally for a while and then it raised new error and exit:
No handlers could be found for logger "kazoo.recipe.watchers"
Exception AttributeError: "SimpleConsumer instance has no attribute '_partitions'" in <bound method SimpleConsumer.__del__ of <pykafka.simpleconsumer.SimpleConsumer at 0x1f2c488 (consumer_group=test2)>> ignored

**Comment by yungchin** [comment]
Thanks @JaneVinWin, that's puzzling actually. I can't quite pin down where in the code that could come from (something that's called by `__del__` apparently). Would you be able to extract more logging/error output, and could you provide it raw, so that all the formatting is preserved?

(I think you can attach plain text files to comments on github, and if that doesn't do there's gist.github.com - thanks!!)

**Comment by emmett9001** [comment]
The most likely origin for this error seems to be [here](https://github.com/Parsely/pykafka/blob/d1573a14933becd4fae24d19f2e487a64f3534da/pykafka/simpleconsumer.py#L384), but there really needs to be more information on this traceback before we can make a clear determination.

**Comment by JaneVinWin** [comment]
@emmett9001  @yungchin , today i upgrade my package to 2.0.2, and it raised new error and the job still hang there but not exit. below is err info:
No handlers could be found for logger "kazoo.client"
Exception in thread Thread-17:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 36, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 333, in _get_children
    result = self._func(children)
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 617, in _topics_changed
    self._rebalance()
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 498, in _rebalance
    self._setup_internal_consumer(list(new_partitions))
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 315, in _setup_internal_consumer
    auto_start=start
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 163, in __init__
    self._discover_offset_manager()
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 264, in _discover_offset_manager
    self._offset_manager = self._cluster.get_offset_manager(self._consumer_group)
  File "/usr/lib/python2.7/site-packages/pykafka/cluster.py", line 301, in get_offset_manager
    raise KafkaException("Socket disconnected during offset manager "
KafkaException: Socket disconnected during offset manager discovery. This can happen when using PyKafka with a Kafka version lower than 0.8.2.

Exception AttributeError: "SimpleConsumer instance has no attribute '_partitions'" in <bound method SimpleConsumer.__del__ of <pykafka.simpleconsumer.SimpleConsumer at 0x357c2d8 (consumer_group=test2)>> ignored
Exception in thread Thread-15:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 208, in checker
    self._rebalance()
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 498, in _rebalance
    self._setup_internal_consumer(list(new_partitions))
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 315, in _setup_internal_consumer
    auto_start=start
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 163, in __init__
    self._discover_offset_manager()
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 264, in _discover_offset_manager
    self._offset_manager = self._cluster.get_offset_manager(self._consumer_group)
  File "/usr/lib/python2.7/site-packages/pykafka/cluster.py", line 301, in get_offset_manager
    raise KafkaException("Socket disconnected during offset manager "
KafkaException: Socket disconnected during offset manager discovery. This can happen when using PyKafka with a Kafka version lower than 0.8.2.

Exception AttributeError: "SimpleConsumer instance has no attribute '_partitions'" in <bound method SimpleConsumer.__del__ of <pykafka.simpleconsumer.SimpleConsumer at 0x3ffbd40 (consumer_group=test2)>> ignored

**Comment by emmett9001** [comment]
@JaneVinWin What version of the Kafka server are you running?

**Comment by JaneVinWin** [comment]
@emmett9001 my Kafka is 0.8.2.1 which qualify the requirement.

**Comment by JaneVinWin** [comment]
@yungchin  @emmett9001 it raised a new err:
Exception in thread Thread-23:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 36, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/kazoo/recipe/watchers.py", line 333, in _get_children
    result = self._func(children)
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 617, in _topics_changed
    self._rebalance()
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 498, in _rebalance
    self._setup_internal_consumer(list(new_partitions))
  File "/usr/lib/python2.7/site-packages/pykafka/balancedconsumer.py", line 315, in _setup_internal_consumer
    auto_start=start
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 163, in __init__
    self._discover_offset_manager()
  File "/usr/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 264, in _discover_offset_manager
    self._offset_manager = self._cluster.get_offset_manager(self._consumer_group)
  File "/usr/lib/python2.7/site-packages/pykafka/cluster.py", line 295, in get_offset_manager
    res = future.get(ConsumerMetadataResponse)
  File "/usr/lib/python2.7/site-packages/pykafka/handlers.py", line 61, in get
    return response_cls(self.response)
  File "/usr/lib/python2.7/site-packages/pykafka/protocol.py", line 907, in __init__
    response = struct_helpers.unpack_from(fmt, buff, 0)
  File "/usr/lib/python2.7/site-packages/pykafka/utils/struct_helpers.py", line 49, in unpack_from
    output = _unpack(fmt, buff, offset, 1)[0]
  File "/usr/lib/python2.7/site-packages/pykafka/utils/struct_helpers.py", line 96, in _unpack
    items.extend(struct.unpack_from('!' + ch, buff, offset))
error: unpack_from requires a buffer of at least 25455 bytes

Exception AttributeError: "SimpleConsumer instance has no attribute '_partitions'" in <bound method SimpleConsumer.__del__ of <pykafka.simpleconsumer.SimpleConsumer at 0x7f163429d1b8 (consumer_group=test2)>> ignored

**Comment by yungchin** [comment]
Thanks very much @JaneVinWin, that's a very helpful traceback. It might take a few more days before I'd get to this, which is why I'm clearing my assignment, but it's definitely still on our plate (we also track this at https://waffle.io/Parsely/pykafka ).

**Comment by yungchin** [comment]
@JaneVinWin just to be sure, could you check if the last traceback you posted was against pykafka 2.0.2 and not an earlier 2.0.2-dev version?  The thing is, I would have thought that this error
```
error: unpack_from requires a buffer of at least 25455 bytes
```
is a manifestation of a bug we very recently fixed in #320, which would explain the apparently corrupted length field in the unpacking. But that was fixed as part of 2.0.2, so if you're running that, then this is some later version.

**Comment by yungchin** [comment]
@emmett9001 the common theme of all the failures here seems to be that exceptions can happen in a zk callback that triggers `_rebalance()`. I wonder if it would make sense to handle that the same way as in the `SimpleConsumer` - so that if any thread (even one spawned by a zk callback) encounters an exception, it stops the whole `BalancedConsumer` and surfaces the exception on the main thread?

**Comment by emmett9001** [comment]
Possibly related to #330

**Comment by emmett9001** [comment]
@yungchin yes, I think that's a good idea - honestly I'm not sure why I didn't set it up that way to begin with. In general I like all of the main pykafka components raising their thread exceptions to the main thread, so I'll take a shot at setting this up.

**Comment by emmett9001** [comment]
@yungchin Pull requested in #331. It doesn't solve this exception, but it at least allows it to be caught by user code.

**Comment by emmett9001** [comment]
I suggest we close this issue. The base issue of thread crashes resulting in zombie threads has been addressed, and there are many different tracebacks of unclear origin. @yungchin @rduplain @JaneVinWin if you'd like to continue tracking this issue, just reopen or comment.

**Comment by JaneVinWin** [comment]
@emmett9001 thanks

**Comment by Lowpower** [comment]
No handlers could be found for logger "pykafka.simpleconsumer"

**Comment by rohitpandey5491** [comment]
Hey @emmett9001: I am trying to write python code for calculating Kafka Consumer Lag, after executing the code, I am getting  no handlers could be found for logger "pykafka.simpleconsumer".
Here is my code:

import pykafka

def fetch_consumer_lag(topic, consumer_group):
    """Get raw lag data for a topic/consumer group.

    :param client: KafkaClient connected to the cluster.
    :type client:  :class:`pykafka.KafkaClient`
    :param topic:  Name of the topic.
    :type topic:  :class:`pykafka.topic.Topic`
    :param consumer_group: Name of the consumer group to fetch lag for.
    :type consumer_groups: :class:`str`
    :returns: dict of {partition_id: (latest_offset, consumer_offset)}
    """
    latest_offsets = fetch_offsets(topic, 'latest')
    # consumer = topic.get_simple_consumer(consumer_group=consumer_group,
    #                                         auto_offset_reset=OffsetType.LATEST,
    #                                         auto_commit_enable=False,
    #                                         auto_commit_interval_ms=1000)


    consumer = topic.get_simple_consumer(consumer_group=consumer_group,
                                         auto_start=False,
                                         auto_offset_reset=OffsetType.LATEST,
                                         consumer_timeout_ms=5000)

    current_offsets = consumer.fetch_offsets()


def kafkaLagCalculator(client):
    """Calculate lag for a topic/consumer group.

    :param client: KafkaClient connected to the cluster.
    :type client:  :class:`pykafka.KafkaClient`
    """
    topic_list = get_all_topics(client)
    consumer_group_topics, consumer_groups = get_consumer_group_for_topic(client, topic_list)

    # Don't auto-create topics.
    for consumer_group_topic in consumer_group_topics:
        if consumer_group_topic not in client.topics:
            raise ValueError('Topic {} does not exist.'.format(consumer_group_topic))

        topic = client.topics[consumer_group_topic]
        threads = [None] * len(consumer_groups)
        i = 0
        try:
            for consumer_group in consumer_groups:
                threads[i] = Thread(target=fetch_consumer_lag, args=(topic, consumer_group))
                threads[i].start()
                i += 1

            for i in range(len(threads)):
                threads[i].join()
        except Exception as e:
            print "ERROR: ", e.message

Any suggestion would be appreciated.
